// @andes/log/dist/src/lib/logger.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mongoose_1 = require("mongoose");
exports.schema = new mongoose_1.Schema({
    level: String,
    type: String,
    start: Date,
    end: Date,
    expiredAt: Date,
    entries: [{
            date: Date,
            action: String,
            usuario: mongoose_1.SchemaTypes.Mixed,
            app: mongoose_1.SchemaTypes.Mixed,
            organizacion: mongoose_1.SchemaTypes.Mixed,
            data: {
                anterior: mongoose_1.SchemaTypes.Mixed,
                valor: mongoose_1.SchemaTypes.Mixed,
            },
            cliente: {
                ip: String,
                userAgent: {
                    isMobile: Boolean,
                    isDesktop: Boolean,
                    isBot: Boolean,
                    browser: String,
                    version: String,
                    os: String,
                    platform: String,
                    source: String
                }
            },
            servidor: {
                ip: String
            },
            error: String
        }]
});
