// @andes/shared/dist/src/lib/operators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var operators_1 = require("rxjs/operators");
var rxjs_1 = require("rxjs");
function notNull() {
    return operators_1.filter(function (user) { return !!user; });
}
exports.notNull = notNull;
function onlyNull() {
    return operators_1.filter(function (user) { return !user; });
}
exports.onlyNull = onlyNull;
function mergeObject() {
    return operators_1.scan(function (acc, curr) { return Object.assign({}, acc, curr); }, {});
}
exports.mergeObject = mergeObject;
function asObject(key, fn) {
    if (fn === void 0) { fn = null; }
    return operators_1.map(function (value) {
        fn = fn || (function (t) { return t; });
        var obj = {};
        obj[key] = value ? fn(value) : null;
        return obj;
    });
}
exports.asObject = asObject;
var replacer = function (key, value) { return value === null ? undefined : value; };
function distincObject() {
    return operators_1.distinctUntilChanged(function (a, b) {
        return JSON.stringify(a, replacer) === JSON.stringify(b, replacer);
    });
}
exports.distincObject = distincObject;
function cache() {
    return rxjs_1.pipe(operators_1.publishReplay(1), operators_1.refCount());
}
exports.cache = cache;
